### ISSUE TITLE
`isinstance(x, cls)` with generic `cls` narrows the type of `x` to the specialized generic type

### ISSUE THREAD

**Issue by CharlesMasson** [microsoft/pyright#11141]
**Describe the bug**

In a class method, `isinstance(..., cls)` narrows the type to the specialized generic type. In all likelihood, other type guards like `issubclass(..., cls)` cause the same behavior.

**Code or Screenshots**

```python
# pyright: strict
from __future__ import annotations

from typing import Self, reveal_type


class A[T]:
    @classmethod
    def cast(cls, x: object) -> Self:
        assert isinstance(x, cls)
        reveal_type(x)  # Type of "x" is "Self@A[T@A]"
        return x


a = A[str]()
reveal_type(A[int].cast(a))  # Type of "A[int].cast(a)" is "A[int]"
reveal_type(A[bool].cast(a))  # Type of "A[bool].cast(a)" is "A[bool]"
```

Pyright does not detect any errors. I'd expect the type of `x` to be `A[Unknown]`, hence a return type error, as is the case when substituting with `isinstance(x, A)`.

**VS Code extension or command-line**

`pyright 1.1.407`

**Comment by erictraut** [comment]
I think pyright is working correctly here. I don't consider this a bug. You will see the same narrowing behavior in other type checkers like mypy.
